File: stack/go/paired_parentheses.go
package stack

// PairedParentheses validates parentheses.
func PairedParentheses(s string) bool {
	balance := 0
	for _, ch := range s {
		switch ch {
		case '(':
			balance++
		case ')':
			balance--
			if balance < 0 {
				return false
			}
		}
	}
	return balance == 0
}

File: stack/js/pairedParentheses.js
const pairedParentheses = (str) => {
  let count = 0;
  
  for (let char of str) {
    if (char === '(') {
      count += 1;
    } else if (char === ')') {
      if (count === 0) {
        return false;
      }
      
      count -= 1;
    }
  }
  
  return count === 0;
};
